interop-testing/src/main/java/io/grpc/testing/integration/RpcBehaviorLoadBalancerProvider.java

```diff
@@ -0,0 +1,172 @@
+    if (rpcBehavior == null) {
+      return ConfigOrError.fromError(
+          Status.INVALID_ARGUMENT.withDescription("no 'rpcBehavior' defined"));
+    }
+    return ConfigOrError.fromConfig(new RpcBehaviorConfig(rpcBehavior));
+  }
+
+  @Override
+  public LoadBalancer newLoadBalancer(Helper helper) {
+    return new RpcBehaviorLoadBalancer(helper,
+        LoadBalancerRegistry.getDefaultRegistry().getProvider("round_robin")
+            .newLoadBalancer(helper));
+  }
+
+  @Override
+  public boolean isAvailable() {
+    return true;
+  }
+
+  @Override
+  public int getPriority() {
+    return 5;
+  }
+
+  @Override
+  public String getPolicyName() {
+    return "test.RpcBehaviorLoadBalancer";
+  }
+
+  static class RpcBehaviorConfig {
+
+    final String rpcBehavior;
+
+    RpcBehaviorConfig(String rpcBehavior) {
+      this.rpcBehavior = rpcBehavior;
+    }
+  }
+
+  /**
+   * Delegates all calls to another LB and wraps the given helper in {@link RpcBehaviorHelper} that
+   * assures that the rpc-behavior metadata header gets added to all calls.
+   */
+  static class RpcBehaviorLoadBalancer extends ForwardingLoadBalancer {
+
+    private final RpcBehaviorHelper helper;
+    private final LoadBalancer delegateLb;
+
+    RpcBehaviorLoadBalancer(Helper helper, LoadBalancer delegateLb) {
+      this.helper = new RpcBehaviorHelper(helper);
+      this.delegateLb = delegateLb;
+    }
+
+    @Override
+    protected LoadBalancer delegate() {
+      return delegateLb;
+    }
+
+    @Override
+    public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+      helper.setRpcBehavior(
+          ((RpcBehaviorConfig) resolvedAddresses.getLoadBalancingPolicyConfig()).rpcBehavior);
+      delegateLb.handleResolvedAddresses(resolvedAddresses);
+    }
+  }
+
+  /**
+   * Wraps the picker that is provided when the balancing change updates with the {@link
+   * RpcBehaviorPicker} that injects the rpc-behavior metadata entry.
+   */
+  static class RpcBehaviorHelper extends ForwardingLoadBalancerHelper {
+
+    private final Helper delegateHelper;
+    private String rpcBehavior;
+
+    RpcBehaviorHelper(Helper delegateHelper) {
+      this.delegateHelper = delegateHelper;
+    }
+
+    void setRpcBehavior(String rpcBehavior) {
+      this.rpcBehavior = rpcBehavior;
+    }
+
+    @Override
+    protected Helper delegate() {
+      return delegateHelper;
+    }
+
+    @Override
+    public void updateBalancingState(@Nonnull ConnectivityState newState,
+        @Nonnull SubchannelPicker newPicker) {
+      delegateHelper.updateBalancingState(newState, new RpcBehaviorPicker(newPicker, rpcBehavior));
+    }
+  }
+
+  /**
+   * Includes the rpc-behavior metadata entry on each subchannel pick.
+   */
+  static class RpcBehaviorPicker extends SubchannelPicker {
+
+    private static final String RPC_BEHAVIOR_HEADER_KEY = "rpc-behavior";
+
+    private final SubchannelPicker delegatePicker;
+    private final String rpcBehavior;
```

Comment:
if i understand correctly, the picker is created in managedChannel syncContext, while pick subchannel is called from created client stream, so volatile would be ideal here, right? cc. @ejona86

Reply:
You are right, except for the volatile part. The picker only needs to worry about synchronizing with itself, and it isn't changing rpcBehavior, so everything is okay.